Regression: build failure on i686-pc-windows-gnu due to multiple definition of `_alloca`

The changes in #575 seem to have broken MSYS2's build of i686-pc-windows-gnu:
```
    = note: D:/M/msys64/mingw32/bin/../lib/gcc/i686-w64-mingw32/13.2.0/../../../../i686-w64-mingw32/bin/ld.exe: D:/M/msys64/mingw32/bin/../lib/gcc/i686-w64-mingw32/13.2.0/libgcc.a(_chkstk.o):(.text+0x0): multiple definition of `_alloca'; C:\_\B\src\MINGW32\build\i686-pc-windows-gnu\stage0-sysroot\lib\rustlib\i686-pc-windows-gnu\lib/libstd-394a669d59370585.dll.a(std_394a669d59370585_dll_d001695.o):(.text+0x0): first defined here
```

Specifically, the latest update renaming `__alloca` to `_alloca` seems to have caused this error.  The version of the changes from commit 67c1c0a71a204c089ddae4aec21ec75aa778c11b (before the last force-push on that PR) work in both i686-pc-windows-gnu and i686-pc-windows-gnullvm.

I'm opening this as a new issue in hopes that that increases visibility.  Hopefully somebody else can also verify that this breaks the 'offical' build of i686-pc-windows-gnu as distributed by rust-lang.org.

_Originally posted by @jeremyd2019 in https://github.com/rust-lang/compiler-builtins/issues/575#issuecomment-2028460311_

Failed also in Rust's CI: https://github.com/rust-lang/rust/pull/123426
`
    = note: C:/a/rust/rust/mingw32/bin/../lib/gcc/i686-w64-mingw32/12.2.0/../../../../i686-w64-mingw32/bin/ld.exe: C:/a/rust/rust/mingw32/bin/../lib/gcc/i686-w64-mingw32/12.2.0/libgcc.a(_chkstk.o):C:/buildroot/src/gcc-12.2.0/libgcc/config/i386/cygwin.S:81: multiple definition of `_alloca'; C:\a\rust\rust\build\i686-pc-windows-gnu\stage0-sysroot\lib\rustlib\i686-pc-windows-gnu\lib/libstd-3e85246dfcced244.dll.a(std_3e85246dfcced244_dll_d001834.o):(.text+0x0): first defined here
`

--- a/src/x86.rs
+++ b/src/x86.rs
@@ -14,8 +14,21 @@ intrinsics! {
         target_env = "gnu",
         not(feature = "no-asm")
     ))]
+    pub unsafe extern "C" fn __chkstk() {
+        core::arch::asm!(
+            "jmp __alloca", // Jump to __alloca since fallthrough may be unreliable"
+            options(noreturn, att_syntax)
+        );
+    }
+
+    #[naked]
+    #[cfg(all(
+        windows,
+        target_env = "gnu",
+        not(feature = "no-asm")
+    ))]
     pub unsafe extern "C" fn _alloca() {
-        // _chkstk and _alloca are the same function
+        // __chkstk and _alloca are the same function
         core::arch::asm!(
             "push   %ecx",
             "cmp    $0x1000,%eax",

